janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLStoreManager.java

```diff
@@ -504,7 +505,8 @@ private void mutateManyLogged(final Map<String, Map<StaticBuffer, KCVMutation>>
 
                 return Iterator.concat(deletions, additions);
             });
-        }));
+        }).collect(Collectors.toList()));
```

Comment:
What if we use `.forEach(builder::addStatement)` instead of `.collect(Collectors.toList())` with `builder.addStatements`?
I'm just thinking that `.collect(Collectors.toList())` creates a new collection in memory whereas `.forEach(builder::addStatement)` doesn't (and also there is not additional iteration needed).

Reply:
I.e. 
```
Iterator.ofAll(mutations.entrySet()).flatMap(tableNameAndMutations -> {
    final String tableName = tableNameAndMutations.getKey();
    final Map<StaticBuffer, KCVMutation> tableMutations = tableNameAndMutations.getValue();

    final CQLKeyColumnValueStore columnValueStore = Option.of(this.openStores.get(tableName))
        .getOrElseThrow(() -> new IllegalStateException("Store cannot be found: " + tableName));
    return Iterator.ofAll(tableMutations.entrySet()).flatMap(keyAndMutations -> {
        final StaticBuffer key = keyAndMutations.getKey();
        final KCVMutation keyMutations = keyAndMutations.getValue();

        final Iterator<BatchableStatement<BoundStatement>> deletions = Iterator.of(commitTime.getDeletionTime(this.times))
            .flatMap(deleteTime -> Iterator.ofAll(keyMutations.getDeletions()).map(deletion -> columnValueStore.deleteColumn(key, deletion, deleteTime)));
        final Iterator<BatchableStatement<BoundStatement>> additions = Iterator.of(commitTime.getAdditionTime(this.times))
            .flatMap(addTime -> Iterator.ofAll(keyMutations.getAdditions()).map(addition -> columnValueStore.insertColumn(key, addition, addTime)));

        return Iterator.concat(deletions, additions);
    });
}).forEach(builder::addStatement);
```

Reply:
@porunov Good idea to make that change, it was in fact my first iteration, and as you say should be more efficient. I only made the change to use `.collect(Collectors.toList())` while trying to prove to myself that that was the cause of the problem with using `addStatements`.